File: cortex_web/apps/web/trainer/conventions.ts
// Canonical parameter-convention bridge (G3 TS port of trainer/conventions.py).
// Deterministic; tolerance-parity gated by conventions.test.ts against the Python
// reference. Reuses the exam engine's normCdf (the shared λ-lapse likelihood);
// adds normPpf (Acklam + one Halley step) for the difficulty-placement constant.
//
//   plan:   z = (s − t)/σ,        (σ, t),  skill = 1/σ
//   engine: z = exp(ℓ)·(s + θ),   (θ, ℓ),  skill = exp(ℓ)
//   bridge: σ = exp(−ℓ),  t = −θ     P(y=1) = λ + (1−2λ)·Φ(z),  λ = 0.025
import { normCdf } from '../engine/mathfns';

export const LAPSE_RATE = 0.025;

export function normPdf(x: number): number {
  return Math.exp(-0.5 * x * x) / Math.sqrt(2 * Math.PI);
}

// Inverse standard normal CDF (Acklam rational approximation + one Halley step).
export function normPpf(p: number): number {
  const a = [-3.969683028665376e1, 2.209460984245205e2, -2.759285104469687e2,
    1.383577518672690e2, -3.066479806614716e1, 2.506628277459239e0];
  const b = [-5.447609879822406e1, 1.615858368580409e2, -1.556989798598866e2,
    6.680131188771972e1, -1.328068155288572e1];
  const c = [-7.784894002430293e-3, -3.223964580411365e-1, -2.400758277161838e0,
    -2.549732539343734e0, 4.374664141464968e0, 2.938163982698783e0];
  const d = [7.784695709041462e-3, 3.224671290700398e-1, 2.445134137142996e0,
    3.754408661907416e0];
  const plow = 0.02425;
  const phigh = 1 - plow;
  let x: number;
  if (p < plow) {
    const q = Math.sqrt(-2 * Math.log(p));
    x = (((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) /
      ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1);
  } else if (p <= phigh) {
    const q = p - 0.5;
    const r = q * q;
    x = (((((a[0] * r + a[1]) * r + a[2]) * r + a[3]) * r + a[4]) * r + a[5]) * q /
      (((((b[0] * r + b[1]) * r + b[2]) * r + b[3]) * r + b[4]) * r + 1);
  } else {
    const q = Math.sqrt(-2 * Math.log(1 - p));
    x = -(((((c[0] * q + c[1]) * q + c[2]) * q + c[3]) * q + c[4]) * q + c[5]) /
      ((((d[0] * q + d[1]) * q + d[2]) * q + d[3]) * q + 1);
  }
  const e = normCdf(x) - p;                       // Halley refinement
  const u = e * Math.sqrt(2 * Math.PI) * Math.exp((x * x) / 2);
  return x - u / (1 + (x * u) / 2);
}

export const WILSON_OPT_ACC = normCdf(1.0);

export function difficultyMultiplier(targetAcc: number, lapse = LAPSE_RATE): number {
  return normPpf((targetAcc - lapse) / (1 - 2 * lapse));
}

export function accuracyAtMultiplier(m: number, lapse = LAPSE_RATE): number {
  return lapse + (1 - 2 * lapse) * normCdf(m);
}

export const SKILL_MODE_MULTIPLIER = difficultyMultiplier(WILSON_OPT_ACC);

// ── parameter conversions ──
export function engineToPlan(theta: number[], ell: number[]): [number[], number[]] {
  return [ell.map((l) => Math.exp(-l)), theta.map((th) => -th)];
}
export function planToEngine(sigma: number[], t: number[]): [number[], number[]] {
  return [t.map((v) => -v), sigma.map((s) => -Math.log(s))];
}

// ── observation model ──
export function pYesPlan(s: number[], sigma: number[], t: number[], lapse = LAPSE_RATE): number[] {
  return s.map((sv, i) => lapse + (1 - 2 * lapse) * normCdf((sv - t[i]) / sigma[i]));
}
export function pYesEngine(s: number[], theta: number[], ell: number[], lapse = LAPSE_RATE): number[] {
  return s.map((sv, i) => lapse + (1 - 2 * lapse) * normCdf(Math.exp(ell[i]) * (sv + theta[i])));
}

// ── skill ↔ AUROC ──
export function aurocFromEll(ell: number[]): number[] {
  return ell.map((l) => normCdf(Math.SQRT2 / Math.sqrt(Math.exp(-2 * l) + 1)));
}
export function aurocFromSigma(sigma: number[]): number[] {
  return sigma.map((s) => normCdf(Math.SQRT2 / Math.sqrt(s * s + 1)));
}
export function sigmaStarFromEllStar(ellStar: number[]): number[] {
  return ellStar.map((e) => Math.exp(-e));
}

File: trainer/conventions.py
"""Canonical parameter-convention bridge (G1 port of trainer_rd bridge_conventions).

Verbatim numerics port of `trainer_rd/training/bridge_conventions.py` — bit-parity
gated by `tests/test_trainer_g1_conventions.py` against golden vectors generated
from the scratch module. Only the module docstring/home changed; every float
operation is identical so the port reproduces the scratch bitwise.

The trainer plan and the eval engine (engine/core.py) parameterize the SAME
probit-lapse observation model two ways:

    plan   : z = (s − t)/σ,        state (σ, t),  skill = 1/σ
    engine : z = exp(ℓ)·(s + θ),   state (θ, ℓ),  skill = exp(ℓ)

so the exact bridge is  σ = exp(−ℓ),  t = −θ  (⚠ SIGN FLIP on the criterion),
and both sides share  P(y=1 | z) = λ + (1 − 2λ)·Φ(z)  with λ = 0.025 locked.

D6: skill-building mode targets the Wilson et al. (2019) optimum (training error
Φ(−1) ≈ 0.1587), giving SKILL_MODE_MULTIPLIER = m(Φ(1)) ≈ 1.0770 under λ = 0.025.
"""
import numpy as np
from scipy.stats import norm

# Must equal engine/core.py LAPSE_RATE (🔒 locked invariant; asserted in
# tests/test_trainer_g1_conventions.py rather than imported, so this module
# stays dependency-light — same contract as the scratch bridge).
LAPSE_RATE = 0.025

# Wilson et al. (2019) optimal training accuracy: error rate Φ(−1) ≈ 0.1587.
WILSON_OPT_ACC = float(norm.cdf(1.0))            # ≈ 0.841345


# ───────────── parameter conversions ─────────────

def engine_to_plan(theta, ell):
    """(θ, ℓ) engine coords → (σ, t) plan coords. Vectorized."""
    theta = np.asarray(theta, dtype=np.float64)
    ell = np.asarray(ell, dtype=np.float64)
    return np.exp(-ell), -theta


def plan_to_engine(sigma, t):
    """(σ, t) plan coords → (θ, ℓ) engine coords. Vectorized; σ > 0."""
    sigma = np.asarray(sigma, dtype=np.float64)
    t = np.asarray(t, dtype=np.float64)
    return -t, -np.log(sigma)


# ───────────── observation model (both conventions) ─────────────

def p_yes_plan(s, sigma, t, lapse=LAPSE_RATE):
    """P(y=1) in plan coords: λ + (1−2λ)·Φ((s − t)/σ)."""
    z = (np.asarray(s, dtype=np.float64) - t) / sigma
    return lapse + (1.0 - 2.0 * lapse) * norm.cdf(z)


def p_yes_engine(s, theta, ell, lapse=LAPSE_RATE):
    """P(y=1) in engine coords: λ + (1−2λ)·Φ(exp(ℓ)·(s + θ))."""
    z = np.exp(np.asarray(ell, dtype=np.float64)) * (np.asarray(s, dtype=np.float64) + theta)
    return lapse + (1.0 - 2.0 * lapse) * norm.cdf(z)


# ───────────── difficulty placement (lapse-corrected) ─────────────

def difficulty_multiplier(target_acc, lapse=LAPSE_RATE):
    """The |s − t|/σ giving expected accuracy `target_acc` under the lapse model.

    m = Φ⁻¹((a − λ)/(1 − 2λ)). Requires λ < a < 1 − λ (the achievable band).
    """
    a = float(target_acc)
    if not (lapse < a < 1.0 - lapse):
        raise ValueError(
            f"target accuracy {a} outside achievable band ({lapse}, {1.0 - lapse})")
    return float(norm.ppf((a - lapse) / (1.0 - 2.0 * lapse)))


def accuracy_at_multiplier(m, lapse=LAPSE_RATE):
    """Inverse of difficulty_multiplier: expected accuracy at |s − t|/σ = m."""
    return float(lapse + (1.0 - 2.0 * lapse) * norm.cdf(float(m)))


# Decision D6: the skill-building mode placement.
SKILL_MODE_MULTIPLIER = difficulty_multiplier(WILSON_OPT_ACC)      # ≈ 1.0770


def skill_mode_multiplier(lapse=LAPSE_RATE, target_acc=WILSON_OPT_ACC):
    """λ-parameterized D6 placement constant (port scaffolding). At port, derive
    placement constants from the instrument's λ through this function rather than
    importing the frozen SKILL_MODE_MULTIPLIER; asymmetric lapse generalizes via
    the same formula with the label-conditional λ."""
    return difficulty_multiplier(target_acc, lapse=lapse)


# ───────────── skill ↔ AUROC ─────────────

def auroc_from_ell(ell):
    """AUROC(ℓ) = Φ(√2/√(exp(−2ℓ)+1)); identical to auroc.auroc_from_l."""
    ell = np.asarray(ell, dtype=np.float64)
    return norm.cdf(np.sqrt(2.0) / np.sqrt(np.exp(-2.0 * ell) + 1.0))


def auroc_from_sigma(sigma):
    """AUROC in plan coords: σ = exp(−ℓ) ⇒ AUROC = Φ(√2/√(σ² + 1))."""
    sigma = np.asarray(sigma, dtype=np.float64)
    return norm.cdf(np.sqrt(2.0) / np.sqrt(sigma * sigma + 1.0))


def sigma_star_from_ell_star(ell_star):
    """Trainer mastery target σ*_k = exp(−ℓ*_k) from a cert cut-score (F14)."""
    return np.exp(-np.asarray(ell_star, dtype=np.float64))
